### ISSUE TITLE
Unexpected behaviour with MultiplayerSpawner after changing scenes via RPC call

### ISSUE THREAD

**Issue by Murchurl** [godotengine/godot#68407]
### Godot version

4.0.beta4

### System information

Windows 11 22H2

### Issue description

I've setup a simple lobby which will only load both players into another scene once a client connects to the host. After the scene is loaded some subscenes are instantiated on the host and added to a child node in the scene. 

What I expect to happen is, the MultiplayerSpawner will spawn the subscenes on the client at the position (0, 0). However it seems that because I am using an rpc call to load the scene for both the host(locally) and client(remotely) the MultiplayerSpawner fails. The code block below is what the debugger returns for the client instance.

```
E 0:00:02:0861   get_node: Node not found: "Main/MultiplayerSpawner" (relative to "/root").
  <C++ Error>    Method/function failed. Returning: nullptr
  <C++ Source>   scene/main/node.cpp:1359 @ get_node()
E 0:00:02:0861   process_simplify_path: Condition "node == nullptr" is true.
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:77 @ process_simplify_path()
E 0:00:02:0861   get_cached_object: ID 2 not found in cache of peer 1.
  <C++ Error>    Condition "!F" is true. Returning: nullptr
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:249 @ get_cached_object()
E 0:00:02:0861   on_spawn_receive: Condition "!spawner" is true. Returning: ERR_DOES_NOT_EXIST
  <C++ Source>   modules/multiplayer/scene_replication_interface.cpp:484 @ on_spawn_receive()
```

The MultiplayerSpawner does sometimes produce the expected results, albeit very infrequently. What baffles me is that I've had a friend test this on their own machine with the same release of Godot and the MultiplayerSpawner seems to work as expected.

I've found that the MultiplayerSpawner behaves as expected if I load the scene manaully for each instance, rather than through an rpc call for both the host(locally) and client(remotely).

### Steps to reproduce

To test loading via RPC
1. Set the project to run with 2 instances
2. Run  Project
3. Select host for one instance and join for the other

To test loading manually
1. Set the project to run with 2 instances
2. Add ```start_game()``` to line 12 and 26 in **lobby.gd**
4. Run  Project
5. Select host for one instance and join for the other

### Minimal reproduction project

[Uploading multiplayer-nodes.zip…]()

**Comment by Roupp** [comment]
I'm experiencing the same problem as I tried to add MultiplayerSpawner to my project, where I use get_tree().change_scene_to_file() method to change between lobby and the gameworld scenes.

In my another demo project the MultiplayerSpawner works fine when everything is contained under one node and I don't change scenes with that method.

I would love a fix or workaround to this problem.

**Comment by Cretezy** [comment]
I'm also experiencing the same issue. I have 2 cases in my testing game:

1. One instance creates a server, another instance connects to the server, they are in a lobby. The server starts the game, which adds the World with the MultiplayerSpawner to both the instances, then spawns the Players (which are synced with the MultiplayerSpawner). This works as intended.
2. One instance creates a server and starts the game, another instance connects to the server. In this case, the client is instructed to add the World (same as the scenario above), however this doesn't work.

I can actually replicate the issue of 2. on 1. by adding World on the client after spawning the Players. This does match the behavior of 2.

From what I can tell: The MultiplayerSpawner doesn't work on the client if it's added after children are already spawned.

**Comment by Cretezy** [comment]
I've found another case where this can happen.

If the client has a `World` scene and the server has a `ServerWorld` scene which is inherited from `World`, the `MultiplayerSpawner` does not work (same error as the OP).

**Comment by baptr** [comment]
I fought with what I think is the same underlying problem as OP (at least) last week.

As I understand it (take with a grain of salt), the replication interface does some work to map node path strings down to IDs to optimize future communication between the peers. Unfortunately, this "simplify_path" operation seems to ~only happens when a peer first connects. If the client hasn't loaded the to-be-synced tree yet when the connection is established, you get:

```
E 0:00:05:0038   get_node: Node not found: "Main/MultiplayerSpawner" (relative to "/root").
  <C++ Error>    Method/function failed. Returning: nullptr
  <C++ Source>   scene/main/node.cpp:1364 @ get_node()
E 0:00:05:0038   process_simplify_path: Condition "node == nullptr" is true.
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:77 @ process_simplify_path()
```

Then, when the server sends the spawn state, the client can't decode the simplified node identifier (since it wasn't able to establish it, and may still not have the node in the tree). Thus, you get these for each to-be-spawned object :

```
E 0:00:05:0038   get_cached_object: ID 2 not found in cache of peer 1.
  <C++ Error>    Condition "!F" is true. Returning: nullptr
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:241 @ get_cached_object()
E 0:00:05:0038   on_spawn_receive: Condition "!spawner" is true. Returning: ERR_DOES_NOT_EXIST
  <C++ Source>   modules/multiplayer/scene_replication_interface.cpp:567 @ on_spawn_receive()
```

In my project, I've hacked around this for now by doing a two-phase start. As long as the client scene loads before the server's MultiplayerSpawner notices the peer has connected, they're able to get the paths simplified on that first sync attempt, and everything is golden after that.

**Comment by baptr** [comment]
Ah, more precisely, it's adding the spawnable nodes in the server's _ready (before the client is present) that triggers the sync.

**Comment by Faless** [comment]
I tried to figure out a good solution for this issue, but it will need to be pushed for 4.1.

The problem relies on the deferred nature of `change_scene_to_*`, and would require the multiplayer API to temporarily stop processing packets, waiting until the scene has changed (which requires extra signals from SceneTree).

I would discourage from using `change_scene_to_*` during a multiplayer session until we address this properly and use a MultiplayerSpawner instead [as explained in this blog post](https://godotengine.org/article/multiplayer-in-godot-4-0-scene-replication/#selecting-levels).

**Comment by trennepohl** [comment]
I'm facing the same issue when a player joins mid match. e.g

Server starts and loads the world
Client 1 `_on_connection_ok` loads the world and notifies server the world has loaded
Server spawns player

Client 2 connects
```
E 0:00:08:0769   get_node: Node not found: "Map/MultiplayerSpawner" (relative to "/root").
  <C++ Error>    Method/function failed. Returning: nullptr
  <C++ Source>   scene/main/node.cpp:1364 @ get_node()
```

The world and the `MultiplayerSpawner` are there, I can see it in the scene tree, but maybe the client didn't load the scene tree fast enough?

**Comment by sygi** [comment]
fwiw, I am getting a similar issue when I'm reparenting a node that has a MultiplayerSynchronizer inside (without changing any scenes): (presumably) the Synchronizer initially tries to use the "old" path. It fixes on its own after a frame.

**Comment by yrucrem** [comment]
Hello everyone!

I get the same error in my client/server setup. Both client and server use the same main scene. The server loads two world scenes on startup wrapped in SubViewport nodes so that I can each assign a World3D and have separate physics. The world scenes contain the MultiplayerSpawner nodes and have each their own MultiplayerAPI instance.

The client will load one of the levels and only after it's loaded it will establish the connection to the respective scenes MultiplayerPeer. The server correctly creates the player object after the connection and obviously tries to sync it to the client, but on the client I get the MultiplayerSpawner error messages:

```
E 0:08:42:0829   get_node: Node not found: "MultiplayerSpawner" (relative to "/root").
  <C++ Error>    Method/function failed. Returning: nullptr
  <C++ Source>   scene/main/node.cpp:1626 @ get_node()
E 0:08:42:0829   process_simplify_path: Condition "node == nullptr" is true.
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:77 @ process_simplify_path()
E 0:08:42:0829   get_cached_object: ID 1 not found in cache of peer 1.
  <C++ Error>    Condition "!F" is true. Returning: nullptr
  <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:241 @ get_cached_object()
E 0:08:42:0829   on_spawn_receive: Condition "!spawner" is true. Returning: ERR_DOES_NOT_EXIST
  <C++ Source>   modules/multiplayer/scene_replication_interface.cpp:571 @ on_spawn_receive()
```

This is how the server tree looks like before the connection:

![image](https://github.com/godotengine/godot/assets/12695328/b86299be-28d3-4fe8-922f-2b7cff6578cc)

This is the server tree after the connection with the spawned player:

![image](https://github.com/godotengine/godot/assets/12695328/28ccda60-a9ab-467c-b777-03ee034b81ac)

And finally the client tree after the connection:

![image](https://github.com/godotengine/godot/assets/12695328/8acd9c01-892b-4296-9658-6003ac652abd)

Any hints for workarounds for this issue or problems with my scene setup would be greatly appreciated.

Best regards,
Patrick

**Comment by xzirox** [comment]
> Hello everyone!
> 
> I get the same error in my client/server setup. Both client and server use the same main scene. The server loads two world scenes on startup wrapped in SubViewport nodes so that I can each assign a World3D and have separate physics. The world scenes contain the MultiplayerSpawner nodes and have each their own MultiplayerAPI instance.
> 
> The client will load one of the levels and only after it's loaded it will establish the connection to the respective scenes MultiplayerPeer. The server correctly creates the player object after the connection and obviously tries to sync it to the client, but on the client I get the MultiplayerSpawner error messages:
> 
> ```
> E 0:08:42:0829   get_node: Node not found: "MultiplayerSpawner" (relative to "/root").
>   <C++ Error>    Method/function failed. Returning: nullptr
>   <C++ Source>   scene/main/node.cpp:1626 @ get_node()
> E 0:08:42:0829   process_simplify_path: Condition "node == nullptr" is true.
>   <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:77 @ process_simplify_path()
> E 0:08:42:0829   get_cached_object: ID 1 not found in cache of peer 1.
>   <C++ Error>    Condition "!F" is true. Returning: nullptr
>   <C++ Source>   modules/multiplayer/scene_cache_interface.cpp:241 @ get_cached_object()
> E 0:08:42:0829   on_spawn_receive: Condition "!spawner" is true. Returning: ERR_DOES_NOT_EXIST
>   <C++ Source>   modules/multiplayer/scene_replication_interface.cpp:571 @ on_spawn_receive()
> ```
> 
> I am using Godot 4.1.2.
> 
> This is how the server tree looks like before the connection:
> 
> ![image](https://private-user-images.githubusercontent.com/12695328/275974152-b86299be-28d3-4fe8-922f-2b7cff6578cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzNjUwNzQsIm5iZiI6MTcwMTM2NDc3NCwicGF0aCI6Ii8xMjY5NTMyOC8yNzU5NzQxNTItYjg2Mjk5YmUtMjhkMy00ZmU4LTkyMmYtMmI3Y2ZmNjU3OGNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDE3MTkzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiNGNkYWE0YWNlOWM4MDJmZDA3MGU1YzA1YzMwY2MxNTY0MDMyYjg5ZWQ4NTJiNDdkNDQ0OThhMWE1ZDhmMTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v-pq5NDDj8Wt7IrNqC5l2DG7e-sPCbpVyhNQ6cVqjPA)
> 
> This is the server tree after the connection with the spawned player:
> 
> ![image](https://private-user-images.githubusercontent.com/12695328/275975300-28ccda60-a9ab-467c-b777-03ee034b81ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzNjUwNzQsIm5iZiI6MTcwMTM2NDc3NCwicGF0aCI6Ii8xMjY5NTMyOC8yNzU5NzUzMDAtMjhjY2RhNjAtYTlhYi00NjdjLWI3NzctMDNlZTAzNGI4MWFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDE3MTkzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyOTQ1ZDQzYTA5ZDk4YzcyNjhhYTg1NGE3ZWFjZTI4YjQwMTQwYzRiZDI2MzIwMjk0Nzg1NTAxNjE2YWFiNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yMFkO77kCere2AveJS-YOOenanzgZquxqEi3RI9uMZ0)
> 
> And finally the client tree after the connection:
> 
> ![image](https://private-user-images.githubusercontent.com/12695328/275975436-8acd9c01-892b-4296-9658-6003ac652abd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzNjUwNzQsIm5iZiI6MTcwMTM2NDc3NCwicGF0aCI6Ii8xMjY5NTMyOC8yNzU5NzU0MzYtOGFjZDljMDEtODkyYi00Mjk2LTk2NTgtNjAwM2FjNjUyYWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDE3MTkzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwMGU2NWIzYjcyODdmZDU1NWIyNWE3Yzc2YTgwOTcwYjM1ODM1YWVjNjI4Y2U1ZTFkMGZkNjM4NmE4ZjhkMGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DO4izLnDmLcRs_4RFO9tR15ttkbLD3vNpEiaODsrlIo)
> 
> Any hints for workarounds for this issue or problems with my scene setup would be greatly appreciated.
> 
> Best regards, Patrick

I too am having this exact issue with trying to build a server that can host multiple different levels where clients can be in different ones.
Did you manage to find a solution to your problem?

**Comment by Faless** [comment]
> Did you manage to find a solution to your problem?

@xzirox see my comment above:

> I would discourage from using `change_scene_to_*` during a multiplayer session until we address this properly and use a MultiplayerSpawner instead [as explained in this blog post](https://godotengine.org/article/multiplayer-in-godot-4-0-scene-replication/#selecting-levels).

**Comment by zedrun00** [comment]
> The problem relies on the deferred nature of change_scene_to_*, and would require the multiplayer API to temporarily stop processing packets, waiting until the scene has changed (which requires extra signals from SceneTree).

I have modified the code for Mini Project and it seems to run normally.

**Comment by vishal-ahirwar** [comment]
I was also  getting same errors with change_scene_to_packed() method

**Comment by g0rski** [comment]
I have the same issue with change_scene_to_file() anyone found some workaround ?

**Comment by Faless** [comment]
> I have the same issue with change_scene_to_file() anyone found some workaround ?

https://github.com/godotengine/godot/issues/68407#issuecomment-1834307813

**Comment by jairodealmeida** [comment]
I heve same problem

multiplayer.multiplayer_peer = server_peer
multiplayer.peer_connected.connect(_add_player_to_game)
multiplayer.peer_disconnected.connect(_del_player)

This connection closes my client window before pass in _del_player method
But when i debug this, and i find my ids, it is correct

![image](https://github.com/user-attachments/assets/2b75f0e1-6dd5-4714-91f9-67fd0a3986ca)

**Comment by Faless** [comment]
This issue is confirmed, **please do not comment to confirm it further**.

The problem relies on the deferred nature of `change_scene_to_*`, and would require the multiplayer API to temporarily stop processing packets or implement some form of scene reconciliation process.

Please do not use `change_scene_to_*` during a multiplayer session until this issue is addressed properly, use a MultiplayerSpawner instead [as explained in this blog post](https://godotengine.org/article/multiplayer-in-godot-4-0-scene-replication/#selecting-levels).

**Comment by jairodealmeida** [comment]
I didn't use change_scene_to_, i use MultiplayerSpawner

**Comment by adamscott** [comment]
> I didn't use change_scene_to_, i use MultiplayerSpawner

The issue is is about using `change_scene_to_` with `MultiplayerSpawner`. If you don't use `change_scene_to_` at all, feel free to create an other issue with a working [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#reporting-bugs).

**Comment by leandrodreamer** [comment]
Hi, I have a very similar problem with an scene MultiplayerSynchronizer not sending data to the newly connected player when a scene is loaded using `change_scene_to_*`, using MultiplayerSpawner as said by @Faless would be the cleanest solution but i found a workaround that seems to work fine, when the player connects and loads the proper scene, the server can remove and then add back the failing MultiplayerSynchronizer node to the scene tree, for me child MultiplayerSpawner still work fine so i have not tested with those (also child MultiplayerSynchronizer of scenes instanced inside those MultiplayerSpawner seem to work fine for me, so not fully sure if is the exact same problem, at least seems related)

`var index := multiplayer_synchronizer.get_index()
var parent : Node = multiplayer_synchronizer.get_parent()
parent.remove_child(multiplayer_synchronizer)
parent.add_child(multiplayer_synchronizer)
parent.move_child(multiplayer_synchronizer, index)`
